### PR TITLE
Fix support for multiple hostnames

### PR DESCRIPTION
**Pull request by pschorf**

### CODE REVIEW

File: waiter/src/waiter/settings.clj [hunk 1]
```diff
@@ -45,6 +45,7 @@
                                           (s/required-key :failed-check-threshold) schema/positive-int}
    (s/required-key :host) schema/non-empty-string
    (s/required-key :hostname) schema/non-empty-string
+   (s/optional-key :alternate-hostnames) [schema/non-empty-string]
```
Comment (shamsimam) [inline waiter/src/waiter/settings.clj]: Lexicographic order :(

File: waiter/test/waiter/core_test.clj [hunk 2]
```diff
@@ -1108,7 +1108,8 @@
       (is (not (waiter-request?-fn {:headers {"host" "waiter-host-1"}})))))
 
   (testing "list hostname config"
-    (let [config {:settings {:hostname ["waiter-host-1" "waiter-host-2"]}}
+    (let [config {:settings {:hostname "waiter-host-1"
+                             :alternate-hostnames ["waiter-host-2"]}}
```
Comment (shamsimam) [inline waiter/test/waiter/core_test.clj]: Lexicographic order :(

Comment (sradack) [reply waiter/test/waiter/core_test.clj]: This is a case where I believe it's reasonable to have `:hostname` first.

File: waiter/src/waiter/settings.clj (hunk 1 above)
Comment (pschorf) [reply waiter/src/waiter/settings.clj]: I think this should be next to hostname, and after it.

File: waiter/test/waiter/core_test.clj (hunk 2 above)
Comment (pschorf) [reply waiter/test/waiter/core_test.clj]: Same here

File: waiter/src/waiter/settings.clj (hunk 1 above)
Comment (pschorf) [reply waiter/src/waiter/settings.clj]: But if it is going to hang up the review, I will just sort it.

File: waiter/test/waiter/token_test.clj [hunk 3]
```diff
@@ -58,29 +58,29 @@
           service-description2 (clojure.walk/stringify-keys
                                  {:cmd "tc2", :cpus 2, :mem 400, :version "d1e2f3", :run-as-user "tu1", :permitted-user "tu3", :token token})
           service-id2 (sd/service-description->service-id service-id-prefix service-description2)
-          waiter-hostname "waiter-hostname.app.example.com"
+          waiter-hostnames #{"waiter-hostname.app.example.com"}
           handle-list-tokens-request (wrap-handler-json-response handle-list-tokens-request)]
 
       (testing "put:unsupported-request-method"
         (let [{:keys [status body]}
               (run-handle-token-request
```
Comment (shamsimam) [inline waiter/test/waiter/token_test.clj]: `run-handle-token-request` parameter should be renamed to `waiter-hostnames` from `waiter-hostname` in the definition above.

**Comment by shamsimam** [comment]
`prepend-waiter-url` is the only function that uses hostname now. Maybe we can do with just a single config entry?

File: waiter/test/waiter/token_test.clj (hunk 3 above)
Comment (pschorf) [reply waiter/test/waiter/token_test.clj]: Done

**Comment by pschorf** [comment]
I agree, reverted back to the old config

File: waiter/src/waiter/core.clj [hunk 4]
```diff
@@ -623,17 +627,18 @@
                                                  router-id async-request-store-atom make-http-request-fn instance-rpc-chan service-id metric-group
                                                  instance reason-map request-properties location response-headers)))
    :prepend-waiter-url (pc/fnk [[:settings port hostname]]
-                         (fn [endpoint-url]
-                           (if (str/blank? endpoint-url)
-                             endpoint-url
-                             (str "http://" hostname ":" port endpoint-url))))
+                         (let [hostname (if (sequential? hostname) (first hostname) hostname)]
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj]: The sequence can be empty, I think we need to enforce in the `settings` schema that the list may not be empty.

Comment (pschorf) [reply waiter/src/waiter/core.clj]: Done

File: waiter/config-full.edn [hunk 5]
```diff
@@ -109,7 +109,8 @@
  :port 9091
 
  ;; Set the DNS name for your Waiter cluster:
- :hostname "waiter.example.com"
```
Comment (sradack) [inline waiter/config-full.edn]: The comment should indicate that a string or non-empty vector of strings are allowed, and that in the event of multiple hostnames, the first entry is the primary hostname.

Comment (pschorf) [reply waiter/config-full.edn]: Done

**Pull request merged**